loki/expression/parser.py

```diff
@@ -0,0 +1,440 @@
+            return math.exp(float([self.rec(par) for par in expr.parameters][0]))
+        return super().map_call(expr)
+
+
+class LokiParser(ParserBase):
+    """
+    String Parser based on [pymbolic](https://github.com/inducer/pymbolic) 's parser.
+
+    The Loki String Parser utilises and extends pymbolic's parser to incorporate
+    Fortran specific syntax and to map pymbolic expressions to Loki expressions, utilising 
+    the mapper ``PymbolicMapper``.
+
+    **Further**, in order to ensure correct ordering of Fortran Statements as documented
+    in ['WD 1539-1 J3/23-007r1 (Draft Fortran 2023)'](https://j3-fortran.org/doc/year/10/10-007.pdf#page=155),
+    pymbolic's parsing logic needed to be slightly adapted.
+
+    Pymbolic references:
+
+    * [GitHub: pymbolic](https://github.com/inducer/pymbolic)
+    * [pymbolic/parser.py](https://github.com/inducer/pymbolic/blob/main/pymbolic/parser.py)
+    * [pymbolic's parser documentation](https://documen.tician.de/pymbolic/utilities.html)
+
+    .. note::
+       **Example:**
+       Reference a declaration node that contains variable "a"
+
+        .. code-block::
+
+            >>> from loki import loki_parse
+            >>> ir
+            Sum((IntLiteral(3, None), Power(IntLiteral(2, None), IntLiteral(4, None))))
+            >>> result = loki_parse('((3+2)/(3-2))**4', evaluate=True)
+            >>> result
+            FloatLiteral('625.0', None)
+    """
+
+    _f_true = intern("f_true")
+    _f_false = intern("f_false")
+    _f_lessequal = intern('_f_lessequal')
+    _f_less = intern('_f_less')
+    _f_greaterequal = intern('_f_greaterequal')
+    _f_greater = intern('_f_greater')
+    _f_equal = intern('_f_equal')
+    _f_notequal = intern('_f_notequal')
+    _f_and = intern("and")
+    _f_or = intern("or")
+    _f_not = intern("not")
+    _f_float = intern("f_float")
+    _f_int = intern("f_int")
+    _f_string = intern("f_string")
+    _f_openbracket = intern("openbracket")
+    _f_closebracket = intern("closebracket")
+
+    lex_table = [
+            (_f_true, pytools.lex.RE(r"\.true\.", re.IGNORECASE)),
+            (_f_false, pytools.lex.RE(r"\.false\.", re.IGNORECASE)),
+            (_f_lessequal, pytools.lex.RE(r"\.le\.", re.IGNORECASE)),
+            (_f_less, pytools.lex.RE(r"\.lt\.", re.IGNORECASE)),
+            (_f_greaterequal, pytools.lex.RE(r"\.ge\.", re.IGNORECASE)),
+            (_f_greater, pytools.lex.RE(r"\.gt\.", re.IGNORECASE)),
+            (_f_equal, pytools.lex.RE(r"\.eq\.", re.IGNORECASE)),
+            (_f_notequal, pytools.lex.RE(r"\.ne\.", re.IGNORECASE)),
+            (_f_and, pytools.lex.RE(r"\.and\.", re.IGNORECASE)),
+            (_f_or, pytools.lex.RE(r"\.or\.", re.IGNORECASE)),
+            (_f_not, pytools.lex.RE(r"\.not\.", re.IGNORECASE)),
+            (_f_float, ("|", pytools.lex.RE(r"[0-9]+\.[0-9]*([eEdD][+-]?[0-9]+)?(_[a-zA-Z]*)", re.IGNORECASE))),
+            (_f_int, pytools.lex.RE(r"[0-9]+?(_[a-zA-Z]*)", re.IGNORECASE)),
```

Comment:
Several limitations of these two patterns (depending how extensive you would like the support to be.

- Currently it mandates a kind specifier (but only the underscore, not the actual specifier), I suppose we wouldn't want to enforce this?
- Also note that kind specifiers can also include numbers or underscores (since it's a parameter name) and can even be only a number:
```
$ cat test_kind.F90
program main
implicit none
integer, parameter :: my_kind = selected_int_kind(5)
integer(kind=selected_int_kind(3)) :: i
i = 1_4 + 1_my_kind
print *,'i=',i
end program
$ gfortran test_kind.F90
$ ./a.out
 i=      2
```
- Real literals can could in principle also be specified in scientific notation, e.g. `1e-8`.

I leave it up to you what level of support you want to provide here, but just as an inspiration, the pattern that fparser uses for signed real literals:
```
>>> from fparser.two.pattern_tools import signed_real_literal_constant
>>> print(signed_real_literal_constant.pattern)
([+-])?\s*((\d+\s*[.]\s*(\d+)?|[.]\s*\d+)\s*([ED]\s*([+-])?\s*\d+)?\s*(_\s*(\d+|[A-Z][\w$]*))?|\d+\s*[ED]\s*([+-])?\s*\d+\s*(_\s*(\d+|[A-Z][\w$]*))?)
```

Reply:
Nice ... I suggest to allow for now: `2.4_jprb`, `2._8`, `2.4e18_my_kind8` and add corresponding tests.